### PR TITLE
Update zktrie with new hash scheme

### PR DESCRIPTION
**Pull request by noel2004**
As title, and prepare for a new version of `v0.6`

- [x] Update module for go-ethereum
- [x] Pass unittests at golang part
- [x] Test trace in go-ethereum
- [x] Update FFI for circuit
- [x] Pass unittests at rust part
- [x] Integration test in super circuit
- [x] Upgrade to `v0.6`

### CODE REVIEW

File: trie/zk_trie_node_test.go [hunk 1]
```diff
@@ -27,29 +27,32 @@ func TestNewNode(t *testing.T) {
...
-		assert.Equal(t, NodeTypeParent, node.Type)
+		node := NewParentNode(NodeTypeBranch_3, k, k)
+		assert.Equal(t, NodeTypeBranch_3, node.Type)
 		assert.Equal(t, k, node.ChildL)
 		assert.Equal(t, k, node.ChildR)
 
 		hash, err := node.NodeHash()
 		assert.NoError(t, err)
-		assert.Equal(t, "11aa224cb5278ea18ffc32f76d8c66ce5caca2bddefb16efb5783c9c8783b0ac", hash.Hex())
+		assert.NotEqual(t, "11aa224cb5278ea18ffc32f76d8c66ce5caca2bddefb16efb5783c9c8783b0ac", hash.Hex())
```
Comment (colinlyguo) [inline trie/zk_trie_node_test.go:54]:
```suggestion
```

File: trie/zk_trie_node_test.go [hunk 2]
```diff
@@ -27,29 +27,32 @@ func TestNewNode(t *testing.T) {
...
 		assert.NoError(t, err)
-		assert.Equal(t, "29311f0403385141efd8ec64fbefc6827a25f553897312a23a403f0681fae600", hash.Hex())
+		assert.Equal(t, "11b483a69ab36a33af13fb95a51a5788fe9936de933595ba30be0a435cba3df1", hash.Hex())
 
 		hash, err = node.ValueHash()
 		assert.NoError(t, err)
-		assert.Equal(t, "13fad1ab739d6ea214e71b03dc3b35e1d7c133b94764b4bc3a45d95bc9a8b12b", hash.Hex())
+		hashFromVp, err := vp[0].Hash()
+		assert.NoError(t, err)
+		assert.Equal(t, hashFromVp.Text(16), hash.Hex())
```
Comment (colinlyguo) [inline trie/zk_trie_node_test.go:42]: better use hardcode hash value?

File: trie/zk_trie_impl.go [hunk 3]
```diff
@@ -665,21 +723,16 @@ func (mt *ZkTrieImpl) walk(nodeHash *zkt.Hash, f func(*Node)) error {
...
-		f(n)
-	case NodeTypeParent:
+	} else {
 		f(n)
 		if err := mt.walk(n.ChildL, f); err != nil {
 			return err
 		}
 		if err := mt.walk(n.ChildR, f); err != nil {
 			return err
 		}
```
Comment (colinlyguo) [inline trie/zk_trie_impl.go:735]: A code style discussion: what if changing `IsTerminal` to `IsEmptyOrLeaf`, and moving other cases outside `IsTerminal`?

File: trie/zk_trie_node.go [hunk 4]
```diff
@@ -25,8 +25,68 @@ const (
...
 )
 
+// DeduceUploadType deduce a new branch type from current branch when one of its child become non-terminal
+func (n NodeType) DeduceUpgradeType(goRight bool) NodeType {
+	if goRight {
+		switch n {
+		case NodeTypeBranch_0:
+			return NodeTypeBranch_1
+		case NodeTypeBranch_1:
+			return n
```
Comment (colinlyguo) [inline trie/zk_trie_node.go:48]:
```suggestion
			return NodeTypeBranch_1
```

File: trie/zk_trie_node.go [hunk 5]
```diff
@@ -25,8 +25,68 @@ const (
...
+			return NodeTypeBranch_3
+		}
+	} else {
+		switch n {
+		case NodeTypeBranch_0:
+			return NodeTypeBranch_2
+		case NodeTypeBranch_1, NodeTypeBranch_3:
+			return NodeTypeBranch_3
+		case NodeTypeBranch_2:
+			return n
```
Comment (colinlyguo) [inline trie/zk_trie_node.go:59]:
```suggestion
			return NodeTypeBranch_2
```

File: trie/zk_trie_impl.go [hunk 6]
```diff
@@ -171,73 +172,86 @@ func (mt *ZkTrieImpl) pushLeaf(newLeaf *Node, oldLeaf *Node, lvl int,
...
 		var newParentNode *Node
 		var newNodeHash *zkt.Hash
+		var isTerminate bool
+		newNodetype := n.Type
 		if path[lvl] { // go right
-			newNodeHash, err = mt.addLeaf(newLeaf, n.ChildR, lvl+1, path, forceUpdate)
-			newParentNode = NewParentNode(n.ChildL, newNodeHash)
+			newNodeHash, isTerminate, err = mt.addLeaf(newLeaf, n.ChildR, lvl+1, path, forceUpdate)
+			if !isTerminate {
+				newNodetype = newNodetype.DeduceUpgradeType(true)
```
Comment (colinlyguo) [inline trie/zk_trie_impl.go:234]:
```suggestion
				newNodetype = newNodetype.DeduceUpgradeType(true) // go right
```

File: trie/zk_trie_impl.go [hunk 7]
```diff
@@ -171,73 +172,86 @@ func (mt *ZkTrieImpl) pushLeaf(newLeaf *Node, oldLeaf *Node, lvl int,
...
+			}
+			newParentNode = NewParentNode(newNodetype, n.ChildL, newNodeHash)
 		} else { // go left
-			newNodeHash, err = mt.addLeaf(newLeaf, n.ChildL, lvl+1, path, forceUpdate)
-			newParentNode = NewParentNode(newNodeHash, n.ChildR)
+			newNodeHash, isTerminate, err = mt.addLeaf(newLeaf, n.ChildL, lvl+1, path, forceUpdate)
+			if !isTerminate {
+				newNodetype = newNodetype.DeduceUpgradeType(false)
```
Comment (colinlyguo) [inline trie/zk_trie_impl.go:240]:
```suggestion
				newNodetype = newNodetype.DeduceUpgradeType(false) // go left
```

File: trie/zk_trie_impl.go [hunk 8]
```diff
@@ -410,45 +444,47 @@ func (mt *ZkTrieImpl) rmAndUpload(path []bool, nodeKey *zkt.Hash, siblings []*zk
...
+	if len(siblings) == 1 { //nolint:gomnd
 		finalRoot = siblings[0]
 		return
 	}
 
+	toUpload := siblings[len(siblings)-1]
+
 	for i := len(siblings) - 2; i >= 0; i-- { //nolint:gomnd
 		if !bytes.Equal(siblings[i][:], zkt.HashZero[:]) {
+			newNodeType := pathTypes[i].DeduceDowngradeType(path[i])
```
Comment (colinlyguo) [inline trie/zk_trie_impl.go:474]:
```suggestion
			newNodeType := pathTypes[i].DeduceDowngradeType(path[i]) // atRight = path[i]
```

File: types/util.go [hunk 9]
```diff
@@ -4,40 +4,48 @@ import (
...
-// reduce 2 fieds into one
-func HashElems(fst, snd *big.Int, elems ...*big.Int) (*Hash, error) {
+// HashElemsWithDomain performs a recursive poseidon hash over the array of ElemBytes, each hash
+// reduce 2 fieds into one, with a specified domain field which would be used in
+// every recursiving call
+func HashElemsWithDomain(domain, fst, snd *big.Int, elems ...*big.Int) (*Hash, error) {
 
 	l := len(elems)
-	baseH, err := hashScheme([]*big.Int{fst, snd})
+	baseH, err := hashScheme([]*big.Int{fst, snd}, domain)
```
Comment (colinlyguo) [inline types/util.go:13]: Would the `domain` field used in every `hashScheme`?

**Review by colinlyguo — APPROVED** [review]

File: trie/zk_trie_node.go [hunk 10]
```diff
@@ -25,8 +25,68 @@ const (
 	// DBEntryTypeRoot indicates the type of a DB entry that indicates the
 	// current Root of a MerkleTree
 	DBEntryTypeRoot NodeType = 3
+
+	NodeTypeLeaf_New  NodeType = 4
```
[the remark this answers is not in the document]
Comment (noel2004) [reply trie/zk_trie_node.go:29]: It is specified in the new doc: https://www.notion.so/scrollzkp/Proposal-for-Poseidon-Hash-Scheme-b6ce299b5bb647ce9de9098f228a1ae9

also it should not be overlapped with the old spec number

[the remark this answers is not in the document]
Comment (colinlyguo) [reply trie/zk_trie_node.go:29]: Oh, I see. btw, thanks for the design doc.

File: src/lib.rs [hunk 11]
```diff
@@ -314,12 +315,19 @@ mod tests {
...
+        domain: *const u8,
+        out: *mut u8,
+    ) -> *const i8 {
         use std::slice;
         let a: [u8; 32] =
             TryFrom::try_from(unsafe { slice::from_raw_parts(a, 32) }).expect("length specified");
         let b: [u8; 32] =
             TryFrom::try_from(unsafe { slice::from_raw_parts(b, 32) }).expect("length specified");
+        let domain: [u8; 32] = TryFrom::try_from(unsafe { slice::from_raw_parts(domain, 32) })
+            .expect("length specified");
```
[the remark this answers is not in the document]
Comment (noel2004) [reply src/lib.rs:330]: [std](https://doc.rust-lang.org/std/index.html)::[slice](https://doc.rust-lang.org/std/slice/index.html)::[from_raw_parts](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#) ensure the returned slice contain expected length

File: trie/zk_trie_node.go (hunk 4 above)
Comment (noel2004) [reply trie/zk_trie_node.go:48]: as mention below

**Pull request merged**